### PR TITLE
Rev blessed contract

### PR DESCRIPTION
**Pull request by birchmd**
## Overview
Introduces the Rev blessed contract into the casper code base.

This is a pretty big PR because of all the new Rholang code coming in. Maybe best to split up the reviewing a little bit. If @pyrocto and @guardbotmk3  want to divide the rholang contracts/tests between the two of you and @KentShikama can take the Scala parts, I think that should make it manageable.

### Does this PR relate to an RChain JIRA issue? 
https://rchain.atlassian.net/browse/RHOL-404

### Complete this checklist before you submit the PR
- [ ] This PR contains no more than 200 lines of code, excluding test code.
- [x] This PR meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards).
- [x] You have someone in mind to assign for review. Make this assignment after submitting the PR.

### Notes
Optional. Add any notes on caveats, approaches you tried that didn't work, or anything else.

### CODE REVIEW

File: casper/src/main/resources/logback.xml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+  <statusListener class="ch.qos.logback.core.status.NopStatusListener" />
+
+  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+    <!-- encoders are  by default assigned the type ch.qos.logback.classic.encoder.PatternLayoutEncoder -->
+    <encoder>
+      <pattern>%d{HH:mm:ss.SSS} [%thread] %-5level %logger{36} - %msg%n</pattern>
+    </encoder>
+  </appender>
+
+  <logger name="coop.rchain.rspace" level="warn" />
```
Comment (KentShikama) [inline casper/src/main/resources/logback.xml]: I was about to do this too 👍

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/BasicWalletSpec.scala [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    runtime)
+  val tuplespace = StoragePrinter.prettyPrint(runtime.space.store)
+
+  "BasicWallet rholang contract" should tests.head in {
+    TestSetUtil.testPassed(tests.head, tuplespace) should be(true)
+  }
+
+  tests.tail.map(test => {
+    it should test in {
+      //TODO: remove when withdraw tests are working properly
```
Comment (KentShikama) [inline casper/src/test/scala/coop/rchain/casper/genesis/contracts/BasicWalletSpec.scala]: I guess we can wait for https://github.com/rchain/rchain/pull/967 ?

File: casper/src/main/rholang/BasicWallet.rho [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+//assumes standard purse API from MakeMint
+contract @"BasicWallet"(@purse, @algorithm, @pk, return) = {
+  new algorithmKnown, cryptoVerifyCh in {
+    match algorithm {
+      "ed25519" => {
+        algorithmKnown!(true) | cryptoVerifyCh!("ed25519Verify")
+      }
+      _ => {
+        algorithmKnown!(false)
```
Comment (guardbotmk3) [inline casper/src/main/rholang/BasicWallet.rho]: It would probably be a bit more concise to use a tuple here and pass Nil as the verify string.

File: casper/src/main/rholang/BasicWallet.rho [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        true => {
+          for (cryptoVerify <- cryptoVerifyCh) {
+            new basicWallet, withdraw, verify, nonceCh in {
+              nonceCh!(-1) |
+              contract verify(@nonce, @sig, return) = {
+                for(@prevNonce <- nonceCh) {
+                  match nonce == (prevNonce + 1) {
+                    true => { 
+                      new result in {
+                        cryptoVerify!(nonce.toByteArray(), sig.hexToBytes(), pk.hexToBytes(), *result) |
```
Comment (guardbotmk3) [inline casper/src/main/rholang/BasicWallet.rho]: I think if we're going to use integers as a nonce, we should have a method that gets a little endian byte array for just the int rather than the process.

File: casper/src/main/rholang/BasicWallet.rho [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                        for(@r <- result) {
+                          match r {
```
Comment (guardbotmk3) [inline casper/src/main/rholang/BasicWallet.rho]: In these cases, I think it would be clearer to use if else.

File: casper/src/main/rholang/BasicWallet.rho (hunk 4 above)
Comment (guardbotmk3) [reply casper/src/main/rholang/BasicWallet.rho]: Also, are we just signing the nonce? That doesn't seem like enough.

File: casper/src/main/rholang/BasicWallet.rho [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+                        }
+                      }
+                    }
+                    _ => { nonceCh!(prevNonce) | return!(false) }
+                  }
+                }
+              } |
+              contract withdraw(@amount, return) = {
+                new destPurseCh, success in {
+                  @[purse, "sprout"]!(*destPurseCh) | for(@destPurse <- destPurseCh) {
```
Comment (guardbotmk3) [inline casper/src/main/rholang/BasicWallet.rho]: It seems this would be simpler if there were a version of sprout that took an amount and a success channel, (or returned a tuple)

File: casper/src/main/rholang/BasicWallet.rho [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+              } |
+              contract @[*basicWallet, "getBalance"](return) = {
+                @[purse, "getBalance"]!(*return)
+              } |
+              contract @[*basicWallet, "deposit"](@amount, @src, success) = {
+                @[purse, "deposit"]!(amount, src, *success)
+              } |
+              contract @[*basicWallet, "withdraw"](@amount, @nonce, @sig, return) = {
+                new verifyResult in {
+                  verify!(nonce, sig, *verifyResult) | for(@r <- verifyResult) {
```
Comment (guardbotmk3) [inline casper/src/main/rholang/BasicWallet.rho]: It seems to me that we should be signing and verifying the amount, and the return channel to avoid pre-emption.

File: casper/src/main/rholang/MakeBrandPair.rho [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+//on a blockchain). However, restricting access is enough to ensure security in
+//some situations. For example one could seal the method on a "purse" object which
+//extracts the tokens inside, thus allowing only those with the unsealer to 
+//take coins out of the purse.
+contract @"MakeBrandPair"(return) = {
+  new seal, unseal, key in {
+    contract seal(@data, return) = {
+      new sealedData in {
+        contract sealedData(@input, return) = {
+          bundle+{
```
Comment (guardbotmk3) [inline casper/src/main/rholang/MakeBrandPair.rho]: I don't know what you're trying to achieve with this bundle. It doesn't seem to accomplish anything.

Comment (guardbotmk3) [reply casper/src/main/rholang/MakeBrandPair.rho]: It seems that what you might want to do instead is to send a list/tuple where the first element is the name, and the second is a bundle0 that uses that name in a linear for.

File: casper/src/main/rholang/MakeMint.rho [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+          contract internalMakePurse(@balance, return) = {
+            //balance must be a name that NonNegativeNumber contract is listening on
+            new thisPurse in {
+              contract @[*thisPurse, "decr"](@amount, success) = { @[balance, "sub"]!(amount, *success) } |
+              contract @[*thisPurse, "getBalance"](return) = { @[balance, "value"]!(*return) } |
+              contract @[*thisPurse, "sprout"](return) = { @[*thisMint, "makePurse"]!(0, *return) } |
+              contract @[*thisPurse, "getDecr"](return) = { @seal!([*thisPurse, "decr"], *return) } |
+              contract @[*thisPurse, "deposit"](@amount, @src, success) = {
+                new result in {
+                  @[src, "getDecr"]!(*result) | for(@sealedSrcDecr <- result) { //call getDecr function in src
```
Comment (guardbotmk3) [inline casper/src/main/rholang/MakeMint.rho]: There's a simpler way to do this without the sealer unsealer. Because we have bundles, we can use unforgeable names in a challenge response protocol.

Mint has a new name as a tag.

getDecr instead of giving a sealed decr would give a new name as a challenge, and an answer channel. If bundle0{*tag | *challenge} is received on the answer channel, decr can be passed.

File: casper/src/main/rholang/MakeMint.rho [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+            new balanceCh in {
+              @["MakeMint", "int2NN"]!(init, *balanceCh) | for(@balance <- balanceCh) {
+                internalMakePurse!(balance, *return)
+              }
+            }
+          } |
+          contract internalMakePurse(@balance, return) = {
+            //balance must be a name that NonNegativeNumber contract is listening on
+            new thisPurse in {
+              contract @[*thisPurse, "decr"](@amount, success) = { @[balance, "sub"]!(amount, *success) } |
```
Comment (guardbotmk3) [inline casper/src/main/rholang/MakeMint.rho]: I think you need `new thisPurse, decr` and then to use `decr` instead of `"decr"` because as it stands, you can just empty a purse on accident, and I don't think that's what you want.

File: casper/src/main/rholang/BasicWallet.rho [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    for(@ak <- algorithmKnown) {
+      match ak {
+        false => { return!([]) }
+        true => {
+          for (cryptoVerify <- cryptoVerifyCh) {
+            new basicWallet, verify, nonceCh in {
+              nonceCh!(-1) |
+              contract verify(@nonce, @sig, return) = {
+                for(@prevNonce <- nonceCh) {
+                  match nonce == (prevNonce + 1) {
```
Comment (pyrocto) [inline casper/src/main/rholang/BasicWallet.rho]: This design makes it easy to mount a denial-of-service attack: just request a nonce and never use it.

File: casper/src/main/rholang/BasicWallet.rho [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+                  nonceCh!(nonce) | return!(nonce)
+                }
+              } |
+              contract @[*basicWallet, "getBalance"](return) = {
+                @[purse, "getBalance"]!(*return)
+              } |
+              contract @[*basicWallet, "deposit"](@amount, @src, success) = {
+                @[purse, "deposit"]!(amount, src, *success)
+              } |
+              contract @[*basicWallet, "withdraw"](@amount, @nonce, @sig, return) = {
```
Comment (pyrocto) [inline casper/src/main/rholang/BasicWallet.rho]: The only thing a wallet should do is verify the signature on a message and then forward it to the underlying purse.  If you think there should be a withdraw method, please put it on the purse itself.

File: casper/src/main/rholang/NonNegativeNumber.rho [hunk 13]
```diff
@@ -0,0 +1,42 @@
+//scalapackage coop.rchain.casper.genesis.contracts
+
+//This is a contract to explicitly implement the int >= 0 behavioural type
+contract @"NonNegativeNumber"(@init, return) = {
+  new this, valueStore in {
+    contract @[*this, "add"](@x, success) = {
+      match x >= 0 {
+        true => {
+          for(@v <- valueStore){
+            valueStore!(v + x) | success!(true)
```
Comment (pyrocto) [inline casper/src/main/rholang/NonNegativeNumber.rho]: At the moment, numbers are signed 32-bit integers.  You have to test for overflow.

File: casper/src/test/rholang/BasicWalletTest.rho [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+      [
+        ["badAlgorithm", []]
+      ]
+    ) |
+  
+    @"BasicWallet"!(purse, "ed25519", pk, "wallet") |
+    for(@[wallet] <- @"wallet") {
+        @"wallet"!([wallet]) |
+        @"getBalance"!(wallet, "walletBalance") |
+        @"deposit"!(wallet, 300, otherPurse, "overdrawDep") |
```
Comment (pyrocto) [inline casper/src/test/rholang/BasicWalletTest.rho]: Add test for overflow.

File: casper/src/test/rholang/NonNegativeNumberTest.rho [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+      contract nnSub(return) = {
+        @[nn, "sub"]!(x, *return)
+      } |
+      return!(*nnSub)
+    }
+  } |
+
+  @"NonNegativeNumber"!(-5, "negBal") |
+  @"NonNegativeNumber"!(15, "properBal") |
+  @"NonNegativeNumber"!(100, "addTest") |
```
Comment (pyrocto) [inline casper/src/test/rholang/NonNegativeNumberTest.rho]: Add test for overflow.

File: casper/src/main/rholang/NonNegativeNumber.rho (hunk 13 above)
Comment (odeac) [reply casper/src/main/rholang/NonNegativeNumber.rho]: It would be nice to only provide arbitrary precision numbers in Rholang, in order to avoid overflows and provide a safer environment for the developers.

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala [hunk 16]
```diff
@@ -0,0 +1,26 @@
...
+    |    //public contract for making empty rev purses
+    |    contract @["Rev", "makePurse"](return) = {
+    |       @[revMint, "makePurse"]!(0, *return)
+    |    } |
+    |
+    |    //basic wallets which exist from genesis
+    |    ${wallets.map(Wallet.rhoCode).mkString(" |\n")}
+    |  }
+    |}
+  """.stripMargin
```
Comment (odeac) [inline casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala]: Can we put these snippets of code in resources instead of embedding them in Scala code?

File: casper/src/main/rholang/NonNegativeNumber.rho [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+          }
+        }
+        _ => { success!(false)  }
+      }
+    } |
+    contract @[*this, "sub"](@x, success) = {
+      match x >= 0 {
```
Comment (odeac) [inline casper/src/main/rholang/NonNegativeNumber.rho]: Using if-then-else instead of match-true-false would be more readable

File: casper/src/main/rholang/BasicWallet.rho [hunk 18]
```diff
@@ -0,0 +1,65 @@
...
+    match algorithm {
+      "ed25519" => {
+        cryptoVerifyCh!(["ed25519Verify"])
+      }
+      _ => {
+        cryptoVerifyCh!([])
+      }
+    } |
+    for(@maybeCryptoVerify <- cryptoVerifyCh) {
+      match maybeCryptoVerify {
```
Comment (odeac) [inline casper/src/main/rholang/BasicWallet.rho]: isn't this a second match on the same condition?  

if at line 11 you just ```return ! ([])``` this match is not needed

File: casper/src/main/rholang/BasicWallet.rho [hunk 19]
```diff
@@ -0,0 +1,65 @@
...
+  new cryptoVerifyCh in {
+    match algorithm {
+      "ed25519" => {
+        cryptoVerifyCh!(["ed25519Verify"])
+      }
+      _ => {
+        cryptoVerifyCh!([])
+      }
+    } |
+    for(@maybeCryptoVerify <- cryptoVerifyCh) {
```
Comment (odeac) [inline casper/src/main/rholang/BasicWallet.rho]: What is the reason behind ```match ... | for (...) {...}```?

File: casper/src/main/rholang/NonNegativeNumber.rho (hunk 13 above)
Comment (pyrocto) [reply casper/src/main/rholang/NonNegativeNumber.rho]: @odeac That's the plan.

File: casper/src/main/rholang/NonNegativeNumber.rho (hunk 17 above)
Comment (pyrocto) [reply casper/src/main/rholang/NonNegativeNumber.rho]: Yes, here and everywhere else.

File: casper/src/main/rholang/BasicWallet.rho (hunk 18 above)
Comment (birchmd) [reply casper/src/main/rholang/BasicWallet.rho]: That's a good point, though it would then also leave the dangling `for(@cryptoVerify <- cryptoVerifyCh)`. Which if fine when we have a garbage collector, but I don't know when we plan on having that.

File: casper/src/main/rholang/BasicWallet.rho [hunk 20]
```diff
@@ -0,0 +1,79 @@
+//scalapackage coop.rchain.rholang.wallet
+
+//assumes standard purse API from MakeMint
+contract @"BasicWallet"(@purse, @algorithm, @pk, return) = {
```
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/rholang/BasicWallet.rho]: Yes. It should always be possible to wrap a purse into a wallet.

File: casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala (hunk 16 above)
Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala]: It uses Scala String Interpolation, so must be in the Scala code.

Comment (birchmd) [reply casper/src/main/scala/coop/rchain/casper/genesis/contracts/Rev.scala]: All pure Rholang is separately given in the `src/main/rholang` and `src/test/rholang` directories

File: casper/src/test/rholang/BasicWalletTest.rho (hunk 14 above)
Comment (birchmd) [reply casper/src/test/rholang/BasicWalletTest.rho]: Done

File: casper/src/test/rholang/NonNegativeNumberTest.rho (hunk 15 above)
Comment (birchmd) [reply casper/src/test/rholang/NonNegativeNumberTest.rho]: Done

**Comment by birchmd** [comment]
@pyrocto @guardbotmk3 Thanks for all the feedback! I think I have addressed all the comments, PTAL

File: casper/src/main/rholang/BasicWallet.rho [hunk 21]
```diff
@@ -0,0 +1,65 @@
...
+        cryptoVerifyCh!(["ed25519Verify"])
+      }
+      _ => {
+        cryptoVerifyCh!([])
+      }
+    } |
+    for(@maybeCryptoVerify <- cryptoVerifyCh) {
+      match maybeCryptoVerify {
+        []             => { return!([]) }
+        [cryptoVerify] => {
```
Comment (goral09) [inline casper/src/main/rholang/BasicWallet.rho]: Can this be written as
```
if(algorithm == "ed25519Verify") {
  …
} else { return!([]) }
```

File: casper/src/main/rholang/BasicWallet.rho [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+                  _ => { nonceCh!(prevNonce) | return!(false) }
+                }
+              }
+            } |
+            contract @[*basicWallet, "getNonce"](return) = {
+              for(@nonce <- nonceCh) {
+                nonceCh!(nonce) | return!(nonce)
+              }
+            } |
+            contract @[*basicWallet, "getBalance"](return) = {
```
Comment (dckc) [inline casper/src/main/rholang/BasicWallet.rho]: Have you considered how this "method dispatch" idiom interacts with ocap patterns such as [revocable forwarder][1]?

In another idiom I've seen, the `"getBalance"` method name is sent along with the args. It's more clear to me how that would compose with revocable forwarder.

I started a [forum discussion on method dispatch idioms][2].


[1]: https://developer.rchain.coop/tutorial/#revocation
[2]: https://forum.rchain.coop/developer/rholang/ocap-happy-method-dispatch-idioms-in-rholang/

File: casper/src/main/rholang/BasicWallet.rho (hunk 21 above)
Comment (birchmd) [reply casper/src/main/rholang/BasicWallet.rho]: But in the future it won't be that simple. The present design easily extends to including other algorithms. In particular we will be include `secp256k1` as soon as you're done adding it to Rholang.

File: casper/src/main/rholang/BasicWallet.rho [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+                          _    => { nonceCh!(prevNonce) | return!(false) }
+                        }
+                      }
+                    }
+                  }
+                  _ => { nonceCh!(prevNonce) | return!(false) }
+                }
+              }
+            } |
+            contract @[*basicWallet, "getNonce"](return) = {
```
[the remark this answers is not in the document]
Comment (birchmd) [reply casper/src/main/rholang/BasicWallet.rho]: Yes, that's right. We are defining a contract on the name `@[*basicWallet, "getNonce"]`. The way this gets used is that when you call `@"BasicWallet"` you get back the unforgable name `basicWallet`, which you can then use to construct the "compound names" where the method contracts can be accessed.

File: casper/src/main/rholang/BasicWallet.rho (hunk 22 above)
Comment (birchmd) [reply casper/src/main/rholang/BasicWallet.rho]: I don't think this idiom is less composable than sending string method names explicitly. In this model capabilities are  given to a name based on the contracts contained at "compound names" you can form with it. The fully featured object has contracts at all the compound names, but you can create a new name with fewer capabilities easily:
```
contract @"createRevocableForwarder"(baseName, allowedMethods, return) = {
  new isAlive, forwarder, addMethod, kill in {
    isAlive!(true) |
    contract addMethod(@method, success) = {
      //there is a minor problem here that you need to handle different
     //number of arguments somehow
      for(@args <- @[*forwarder, method]; _ <- isAlive) {
        @[*baseName, method]!(args) | 
        success!(true) |
        isAlive!(true) | 
        addMethod!(method, Nil)
      }
    } |
    contract kill(success) = {
      for(_ <- isAlive) {
        //Do not return any data to the `isAlive` channel, so the 
       //join above will no longer trigger.
        success!(true)
      }
    } |
    //initialize
    new methodsAdded in {
      @["LinkedList", "foreach"]!(allowedMethods, *addMethod, *methodsAdded) |
      for(_ <- methodsAdded) {
        return!([*forwarder, *kill])
      }
    }
  }
}
```
The contract above creates a revocable forwarder with any number of capabilities, which are specified at construction of the forwarder. There is a minor hiccup in that if the methods you wanted to forward took multiple arguments then `args` would need to be a list and you would need to deconstruct that list and pass its elements on to the underlying method.

Comment (dckc) [reply casper/src/main/rholang/BasicWallet.rho]: Thanks for spelling that out in detail.

It seems a little awkward, but maybe that's mostly a matter of what I'm familiar with.
